parlai/core/worlds.py

```diff
@@ -1245,6 +1252,127 @@ def shutdown(self):
         self.inner_world.shutdown()
 
 
+class QBatchWorld(BatchWorld):
+    """
+    BatchWorld Where batch_act is Fake!!
+
+    World just queues up the batch!
+    """
+
+    def __init__(
+        self,
+        opt: Opt,
+        world: World,
+        produce_queue: Queue,
+        consume_queue: Queue,
+        produce_cond: Condition,
+        consume_cond: Condition,
+    ):
+        super().__init__(opt, world)
+        self.produce_queue = produce_queue
+        self.consume_queue = consume_queue
+        self.produce_cond = produce_cond
+        self.consume_cond = consume_cond
+
+    def parley(self):
+        """
+        Produce a batch observation.
+
+        Consume a batch act.
+        """
+        if hasattr(self.world, 'parley_init'):
+            for w in self.worlds:
+                w.parley_init()
+        teacher_idx = 0
+        agent_idx = 1
+        with self.produce_cond:
+            self.produce_cond.wait_for(self.produce_queue.empty)
+            batch_act = self.batch_act(teacher_idx, None)
+            batch_obs = self.batch_observe(agent_idx, batch_act, teacher_idx)
+            self.produce_queue.put(batch_obs)
+        with self.consume_cond:
+            self.consume_cond.wait_for(self.consume_queue.full)
+            batch_act = self.consume_queue.get()
+            for other_idx in range(len(self.world.get_agents())):
+                self.batch_observe(other_idx, batch_act, agent_idx)
+        self.update_counters()
+
+
+def run_q_world(opt, world, mp_structs):
+    world = QBatchWorld(
+        opt,
+        world,
+        mp_structs['produce_queue'],
+        mp_structs['consume_queue'],
+        mp_structs['produce_cond'],
+        mp_structs['consume_cond'],
+    )
+    while not world.epoch_done():
+        world.parley()
+
+
+class QueueWorld(BatchWorld):
+    """
+    QueueWorld creates subprocess worlds.
+
+    They create their own batches.
+    """
+
+    def __init__(self, opt: Opt, world: World):
+        World.__init__(self, opt)
+        # QueueWorld init
+        self.world: World = world
+        self.processes: List[Process] = []
+        self.structs: List[Dict[str, Union[Queue, Condition]]] = []
+        self.worlds: List[World] = []
+        for _ in range(opt['numworkers']):
+            produce_cond = Condition()
+            consume_cond = Condition()
+            produce_queue = Queue(maxsize=1)
+            consume_queue = Queue(maxsize=1)
+            shared = world.share()
+            subworld = shared['world_class'](opt, None, shared)
```

Comment:
is the none the agent?!

Reply:
i think so; i took this logic directly from `BatchWorld`

Reply:
Okay I need to understand this cuz it scares me

Reply:
okay I see. It has two mutually exclusive arguments, shared and agents. Maybe we can drop the none and do `(opt, shared=shared)` to make it slightly more clear